src/Common/ConcurrencyControl.h

```diff
@@ -0,0 +1,207 @@
+#pragma once
+
+#include <base/types.h>
+#include <boost/core/noncopyable.hpp>
+#include <mutex>
+#include <memory>
+#include <list>
+#include <condition_variable>
+
+/*
+ * Controls how many threads can be allocated for a query (or another activity).
+ * There is a limited amount of slots for threads. It can be set with `setMaxConcurrency(limit)`.
+ *
+ * Lifecycle of a slot: free -> granted -> acquired -> free.
+ * free: slot is available to be allocated by any query.
+ * granted: slot is allocated by specific query, but not yet acquired by any thread.
+ * acquired: slot is allocated by specific query and acquired by a thread.
+ *
+ * USAGE:
+ *   1. Create an allocation for a query:
+ *      `auto slots = ConcurrencyControl::instance().allocate(min, max);`
+ *      It will allocate at least `min` and at most `max` slots.
+ *      Note that `min` slots are granted immediately, but other `max - min` may be granted later.
+ *   2. For every thread a slot has to be acquired from that allocation:
+ *      `while (auto slot = slots->tryAcquire()) createYourThread([slot = std::move(slot)] { ... });`
+ *      This snippet can be used at query startup and for upscaling later.
+ * (both functions are non-blocking)
+ *
+ * Released slots are distributed between waiting allocations in a round-robin manner to provide fairness.
+ * Oversubscription is possible: total amount of allocated slots can exceed `setMaxConcurrency(limit)`
+ * because `min` amount of slots is allocated for each query unconditionally.
+ */
+class ConcurrencyControl : boost::noncopyable
+{
+public:
+    struct Allocation;
+    using AllocationPtr = std::shared_ptr<Allocation>;
+    using Slots = UInt64;
+    using Waiters = std::list<Allocation *>;
+
+    // Scoped guard for acquired slot, see Allocation::tryAcquire()
+    struct Slot : boost::noncopyable
+    {
+        ~Slot()
+        {
+            allocation->release();
+        }
+
+    private:
+        friend struct Allocation; // for ctor
+
+        explicit Slot(AllocationPtr && allocation_)
+            : allocation(std::move(allocation_))
+        {}
+
+        AllocationPtr allocation;
+    };
+
+    // FIXME: have to be unique_ptr, but ThreadFromGlobalPool does not support move semantics yet
+    using SlotPtr = std::shared_ptr<Slot>;
+
+    // Manages group of slots for a single query, see ConcurrencyControl::allocate(min, max)
+    struct Allocation : std::enable_shared_from_this<Allocation>, boost::noncopyable
+    {
+        ~Allocation()
+        {
+            if (released == limit) // also equal to `allocated`: everything is already released
+                return;
+
+            std::unique_lock lock{parent.mutex};
+            parent.cur_concurrency -= allocated - released;
```

Comment:
Is it really needed to use parent.mutex directly?
Why not make ConcurrencyControl interface complete?

Reply:
Just to keep things as simple as possible, they are tightly bound anyways now. Probably can be improved later if someone needs more complicated logic without Allocation.

Reply:
Thanks for handling it